### ISSUE TITLE
Unable to integrate PaperJS with externally imported JSDOM

### ISSUE THREAD

**Issue by elliotbonneville** [paperjs/paper.js#1048]
For the past several days, I've been trying to integrate PaperJS with my own `import` of JSDOM for automated testing of event handling in an ES6 context (I'm running Webpack with Babel for transpilation). I can't rely on the internal `require` of JSDOM that Paper does in the absence of `self.window`, because it throws an error when `HTMLCanvasElementImpl.dispatchEvent` is passed an instance of `KeyboardEvent` or `MouseEvent` instantiated from the externally `import`ed JSDOM:
```
TypeError: Argument to dispatchEvent must be an Event
      at HTMLCanvasElementImpl.dispatchEvent (node_modules/paper/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:83:13)
      at HTMLCanvasElementImpl.dispatchEvent (node_modules/paper/node_modules/jsdom/lib/jsdom/living/nodes/HTMLElement-impl.js:30:27)
      at HTMLCanvasElement.dispatchEvent (node_modules/paper/node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:71:59)
      at Context.<anonymous> (spec.js:34:36) 
```

To my understanding this is caused because JSDOM relies on an `instanceof` check against its `EventImpl` class, which fails when passed an instance of `EventImpl` from an ES6-style `import` of JSDOM: https://github.com/tmpvar/jsdom/blob/master/lib/jsdom/living/events/EventTarget-impl.js#L77

My solution to this was to populate the `self` variable that Paper checks for when initializing, with the `window` object provided by my `import` of JSDOM. This prevents it from relying on its internal JSDOM `require`, which stops `dispatchEvent` from complaining. Yay, I can trigger events to my heart's content... when all my modules execute in the order I want.

Unfortunately, the ES6 spec states that all `import`ed modules execute asynchronously. Babel seems  to conform to this, so I can't guarantee that `self` will be defined by the time the IIFE that wraps PaperJS executes. This means that sometimes PaperJS does end up relying on its internal `require` of JSDOM.

I believe that a fix for this would be to export that init function instead of executing it, when running in an ES6 context. This would allow me to call it myself after `self` is populated. Is this something that you would consider, or could you suggest an alternate solution?

**Comment by lehni** [comment]
Did you close this on purpose? Did you find a fix?

**Comment by elliotbonneville** [comment]
Hi Jurg,

Yeah, I did -- turned out to be quite simple in the end. Instead of importing my own instance of JSDOM to get access to the `MouseEvent` class, I used the `MouseEvent` on the JSDOM `window` object that Paper exposes. That fixed the `Argument to dispatchEvent must be an Event` error.

As a side note (perhaps worthy of another thread...), there were a couple of other properties I had to set on the `document` to get Paper to accept `MouseEvent`s:

```
const html = document.documentElement;
const body = document.body;

html.scrollLeft = 0;
html.scrollTop = 0;
html.clientLeft = 0;
html.clientTop = 0;

body.scrollLeft = 0;
body.scrollTop = 0;
```

I also had to pass in the `clientX` and `clientY` properties to `MouseEvent`s:

```
paper.project.view.element.dispatchEvent(new window.MouseEvent('mousedown', {
    clientX: 10,
    clientY: 10,
}));
```

Thanks for following up on this.
